hibernate-core/src/main/java/org/hibernate/query/hql/internal/SemanticQueryBuilder.java

```diff
@@ -2102,7 +2104,22 @@ else if ( ( possibleEnumValues = getPossibleEnumValues( rightExpressionContext )
 					break;
 				}
 				left = (SqmExpression<?>) leftExpressionContext.accept( this );
-				right = (SqmExpression<?>) rightExpressionContext.accept( this );
+				if ( left instanceof AnyDiscriminatorSqmPath ) {
+					final SqmPath anyDiscriminatorSqmPath = ( (AnyDiscriminatorSqmPath) left ).getLhs();
+					final EntityDomainType<Object> entityWithDiscriminator = creationContext.getJpaMetamodel()
+							.entity( anyDiscriminatorSqmPath.findRoot().getNavigablePath().getLocalName() );
+					final EntityDomainType<Object> entityDiscriminatorValue = creationContext.getJpaMetamodel()
+							.resolveHqlEntityReference( rightExpressionContext.getText() );
+					right = new SqmAnyDiscriminatorValue<>(
+							entityWithDiscriminator,
+							anyDiscriminatorSqmPath.getNodeType().getPathName(),
+							entityDiscriminatorValue,
+							creationContext.getNodeBuilder()
+					);
+				}
+				else {
+					right = (SqmExpression<?>) rightExpressionContext.accept( this );
+				}
```

Comment:
To handle that the discriminator value is also on the other hand side.

```suggestion
			final SqmExpression<?> l = (SqmExpression<?>) leftExpressionContext.accept( this );
			final SqmExpression<?> r = (SqmExpression<?>) rightExpressionContext.accept( this );

			if ( l instanceof AnyDiscriminatorSqmPath && r instanceof SqmLiteralEntityType ) {
				left = l;
				right = createDiscriminatorValue( left, r );
			}
			else if ( r instanceof AnyDiscriminatorSqmPath && l instanceof SqmLiteralEntityType ) {
				left = createDiscriminatorValue( r, l );
				right = r;
			}
			else {
				left = l;
				right = r;
			}
```

`createDiscriminatorValue` is essentially:

```
	final SqmPath anyDiscriminatorSqmPath = ( (AnyDiscriminatorSqmPath) left ).getLhs();
	final EntityDomainType<Object> entityWithDiscriminator = creationContext.getJpaMetamodel()
			.entity( anyDiscriminatorSqmPath.findRoot().getNavigablePath().getLocalName() );
	final EntityDomainType<Object> entityDiscriminatorValue = creationContext.getJpaMetamodel()
			.resolveHqlEntityReference( rightExpressionContext.getText() );
	return new SqmAnyDiscriminatorValue<>(
			entityWithDiscriminator,
			anyDiscriminatorSqmPath.getNodeType().getPathName(),
			entityDiscriminatorValue,
			creationContext.getNodeBuilder()
	);
```

Reply:
done